protobuf/lib/src/protobuf/mixins/well_known.dart

```diff
@@ -65,4 +162,573 @@ abstract class TimestampMixin {
     target.seconds = Int64(micros ~/ Duration.microsecondsPerSecond);
     target.nanos = (micros % Duration.microsecondsPerSecond).toInt() * 1000;
   }
+
+  static String _twoDigits(int n) {
+    if (n >= 10) return "${n}";
+    return "0${n}";
+  }
+
+  static final DateTime _minTimestamp = DateTime.utc(1);
+  static final DateTime _maxTimestamp = DateTime.utc(9999, 13, 31, 23, 59, 59);
+
+  // From google/protobuf/timestamp.proto:
+  // # JSON Mapping
+  //
+  // In JSON format, the Timestamp type is encoded as a string in the
+  // [RFC 3339](https://www.ietf.org/rfc/rfc3339.txt) format. That is, the
+  // format is "{year}-{month}-{day}T{hour}:{min}:{sec}[.{frac_sec}]Z"
+  // where {year} is always expressed using four digits while {month}, {day},
+  // {hour}, {min}, and {sec} are zero-padded to two digits each. The fractional
+  // seconds, which can go up to 9 digits (i.e. up to 1 nanosecond resolution),
+  // are optional. The "Z" suffix indicates the timezone ("UTC"); the timezone
+  // is required. A proto3 JSON serializer should always use UTC (as indicated by
+  // "Z") when printing the Timestamp type and a proto3 JSON parser should be
+  // able to accept both UTC and other timezones (as indicated by an offset).
+  //
+  // For example, "2017-01-15T01:30:15.01Z" encodes 15.01 seconds past
+  // 01:30 UTC on January 15, 2017.
+  static Object toProto3JsonHelper(
+      GeneratedMessage message, TypeRegistry typeRegistry) {
+    TimestampMixin timestamp = message as TimestampMixin;
+    DateTime dateTime = timestamp.toDateTime();
+
+    if (timestamp.nanos < 0) {
+      throw ArgumentError(
+          'Timestamp with negative `nanos`: ${timestamp.nanos}');
+    }
+    if (timestamp.nanos > 999999999) {
+      throw ArgumentError(
+          'Timestamp with `nanos` out of range: ${timestamp.nanos}');
+    }
+    if (dateTime.isBefore(_minTimestamp) || dateTime.isAfter(_maxTimestamp)) {
+      throw ArgumentError('Timestamp Must be from 0001-01-01T00:00:00Z to '
+          '9999-12-31T23:59:59Z inclusive. Was: ${dateTime.toIso8601String()}');
+    }
+
+    // Because [DateTime] doesn't have nano-second precision, we cannot use
+    // dateTime.toIso8601String().
+    String y = '${dateTime.year}'.padLeft(4, '0');
+    String m = _twoDigits(dateTime.month);
+    String d = _twoDigits(dateTime.day);
+    String h = _twoDigits(dateTime.hour);
+    String min = _twoDigits(dateTime.minute);
+    String sec = _twoDigits(dateTime.second);
+    String secFrac = timestamp.nanos
+        .toString()
+        .padLeft(9, '0')
+        .replaceFirst(finalGroupsOfThreeZeroes, '');
+    if (secFrac != '') secFrac = '.$secFrac';
+    return "$y-$m-${d}T$h:$min:$sec${secFrac}Z";
+  }
+
+  static void fromProto3JsonHelper(GeneratedMessage message, Object json,
+      TypeRegistry typeRegistry, JsonParsingContext context) {
+    if (json is String) {
+      Match fracSecsMatch = RegExp(r'\.(\d*)').firstMatch(json);
```

Comment:
May want `RegExp(r'\.(\d+)')` to ensure that you don't accept `.` with no digit after (unless you want to, but I doubt it's a valid format).

Reply:
Good catch. The spec indeed says at least one digit